### PR TITLE
[src] Add a frame offset setting to online2 feature pipelines.

### PR DESCRIPTION
**Pull request by dogancan**
Relates to #2938.

### CODE REVIEW

File: src/online2/online-feature-pipeline.h [hunk 1]
```diff
@@ -164,9 +164,17 @@ class OnlineFeaturePipeline: public OnlineFeatureInterface {
   virtual int32 NumFramesReady() const;
   virtual void GetFrame(int32 frame, VectorBase<BaseFloat> *feat);
 
+  /// Set the frame offset value. Frame offset is initialized to 0 when the
+  /// feature pipeline is constructed and stays as 0 unless this method is
+  /// called. This method is useful when we want to reset the decoder state,
+  /// i.e. call decoder.InitDecoding(), but we want to keep using the same
+  /// feature pipeline, e.g. in case of an endpoint. The frame offset affects
+  /// the behavior of IsLastFrame, NumFramesReady and GetFrame methods.
```
Comment (danpovey) [inline src/online2/online-feature-pipeline.h]: Thanks.  Please use the () when referring to functions in documentation.

Comment (dogancan) [reply src/online2/online-feature-pipeline.h]: Done. Also added a check to make sure `0 <= frame_offset`.

**Comment by danpovey** [comment]
Thanks.  @danijel3 would you mind merging this into your PR #2938  so we can have the benefit of that testing?  (Not that I anticipate problems).  I'll leave this one open though, till we merge that.

**Comment by danijel3** [comment]
Okay, I'll be monitoring this thread, just like the other one.

**Comment by dogancan** [comment]
@danpovey What do you think about this new design? I didn't add a new decodable for applying the frame offset but instead added a frame offset variable to online looped decodable objects.

With this design, the workflow for multiple utterances in a single feature pipeline will be something like:

```C++
int32 frame_offset = 0;

while(more_chunks) {
  feature_pipeline.AcceptWaveform(samp_freq, chunk);
  if (last_chunk) {
    feature_pipeline.InputFinished();
  }
  if (silence_weighting.Active()) {
    silence_weighting.ComputeCurrentTraceback(decoder.Decoder());
    silence_weighting.GetDeltaWeights(feature_pipeline.NumFramesReady(), &delta_weights);
    feature_pipeline.UpdateFrameWeights(delta_weights, frame_offset * frame_subsampling_factor);
  }
  decoder.AdvanceDecoding();
  if (decoder.EndpointDetected(endpoint_opts)) {
    // output current utterance
    frame_offset += decoder.NumFramesDecoded();
    decoder.InitDecoding(frame_offset);
  }
}
decoder.FinalizeDecoding();
// output last utterance
```

**Comment by danpovey** [comment]
I could be OK with that in principle; some things  are not clear to me from
that snippet of code, but perhaps you could make a PR with just the key
parts / interface changes and I could review that?  or just do the whole
thing-- chances are I will be fine with it.

**Comment by danijel3** [comment]
@dogancan Hey, can I do anything to help resolve this PR?

**Comment by danpovey** [comment]
I think this was already done via a different PR.

**Comment by kkm000** [comment]
Merged with #2938

**Pull request closed without merging**